lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -778,21 +778,52 @@ impl TryFrom<BitmapEvent> for CGOBitmap {
             data_cap: 0,
         };
 
-        // e.decompress consumes e, so we need to call it separately, after populating the fields
-        // above.
-        let mut data = e.decompress()?;
-        res.data_ptr = data.as_mut_ptr();
-        res.data_len = data.len();
-        res.data_cap = data.capacity();
+        let w: u16 = e.width;
+        let h: u16 = e.height;
+
+        let mut encoded = Vec::with_capacity(8192);
+        encode_png(&mut encoded, w, h, e.decompress()?);
+
+        res.data_ptr = encoded.as_mut_ptr();
+        res.data_len = encoded.len();
+        res.data_cap = encoded.capacity();
 
         // Prevent the data field from being freed while Go handles it.
         // It will be dropped once CGOBitmap is dropped (see below).
-        mem::forget(data);
+        mem::forget(encoded);
 
         Ok(res)
     }
 }
 
+pub fn encode_png(dest: &mut Vec<u8>, width: u16, height: u16, mut data: Vec<u8>) {
+    swap(&mut data);
+
+    let mut encoder = png::Encoder::new(dest, width as u32, height as u32);
+    encoder.set_compression(png::Compression::Fast);
+    encoder.set_color(png::ColorType::Rgba);
+
+    let mut writer = encoder.write_header().unwrap();
+    writer.write_image_data(&data).unwrap();
+    writer.finish().unwrap();
+}
+
+/// Convert BGRA to RGBA. It's likely due to Windows using uint32 values for
+/// pixels (ARGB) and encoding them as big endian. The image.RGBA type uses
+/// a byte slice with 4-byte segments representing pixels (RGBA).
+///
+/// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/8ab64b94-59cb-43f4-97ca-79613838e0bd
+///
+/// Also, always force Alpha value to 100% (opaque). On some Windows
+/// versions (e.g. Windows 10) it's sent as 0% after decompression for some reason.
+fn swap(data: &mut Vec<u8>) {
+    let mut i = 0;
+    while i < data.len() {
+        (data[i], data[i + 2], data[i + 3]) = (data[i + 2], data[i], 255);
+        i += 4;
```

Comment:
```suggestion
    data.chunks_exact_mut(4).for_each(|chunk| {
        chunk.swap(0, 2);
        chunk[3] = 255;
    });
```
we check `i < data.len()` but index ahead to `i+3`.

I assume it's always divisible by 4 since this is BGRA data, but this should also be a bit faster

Reply:
Thanks for suggestion! It's always better to use functions that keep track of indexes rather than doing it manually.